atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -2527,38 +2529,99 @@ private void logCommitLockTenureExceeded(
                 stackTraceSnapshot);
     }
 
-    protected Set<LockDescriptor> getLocksForWrites() {
-        Set<LockDescriptor> result = new HashSet<>();
+    protected LocksAndMetadata getLocksAndMetadataForWrites() {
+        ImmutableSet.Builder<LockDescriptor> lockDescriptorBuilder = ImmutableSet.builder();
+        ImmutableMap.Builder<LockDescriptor, ChangeMetadata> lockDescriptorToChangeMetadataBuilder =
+                ImmutableMap.builder();
         long cellLockCount = 0L;
         long rowLockCount = 0L;
-        for (TableReference tableRef : writesByTable.keySet()) {
+        for (TableReference tableRef : localWriteBuffer.getLocalWrites().keySet()) {
             ConflictHandler conflictHandler = getConflictHandlerForTable(tableRef);
             if (conflictHandler.lockCellsForConflicts()) {
-                NavigableSet<Cell> cellsToLock = getLocalWrites(tableRef).keySet();
-                for (Cell cell : cellsToLock) {
-                    result.add(AtlasCellLockDescriptor.of(
-                            tableRef.getQualifiedName(), cell.getRowName(), cell.getColumnName()));
-                }
-                cellLockCount += cellsToLock.size();
+                cellLockCount +=
+                        collectCellLocks(lockDescriptorBuilder, lockDescriptorToChangeMetadataBuilder, tableRef);
             }
-
             if (conflictHandler.lockRowsForConflicts()) {
-                Cell lastCell = null;
-                for (Cell cell : getLocalWrites(tableRef).keySet()) {
-                    if (lastCell == null || !Arrays.equals(lastCell.getRowName(), cell.getRowName())) {
-                        result.add(AtlasRowLockDescriptor.of(tableRef.getQualifiedName(), cell.getRowName()));
-                        rowLockCount++;
-                    }
-                    lastCell = cell;
-                }
+                rowLockCount += collectRowLocks(lockDescriptorBuilder, lockDescriptorToChangeMetadataBuilder, tableRef);
             }
         }
-        result.add(AtlasRowLockDescriptor.of(
+        lockDescriptorBuilder.add(AtlasRowLockDescriptor.of(
                 TransactionConstants.TRANSACTION_TABLE.getQualifiedName(),
                 TransactionConstants.getValueForTimestamp(getStartTimestamp())));
         cellCommitLocksRequested = cellLockCount;
         rowCommitLocksRequested = rowLockCount + 1;
-        return result;
+        Map<LockDescriptor, ChangeMetadata> lockDescriptorToChangeMetadata =
```

Comment:
Considering we're touching this path quite aggressively, and it appears we do not have the best unit tests for this class, could we write some additional tests for this in particular? I know we have `metadataIsTransferredToCellLocks` and other various tests, but I'd like something a bit more extensive here (maybe some random/proprety-based-testing). We have really poor testing over this code path presently :(

Reply:
Gonna try to do some extensive random testing

Reply:
I have also explicitly added the regular locks that we expect to be present in the request to the `metadata` tests so we get additional coverage